### PR TITLE
Added Exos vlan resource module

### PR DESCRIPTION
**Pull request by JayalakshmiV**
##### SUMMARY

Resource module for exos_vlan, with supported operations and integration tests.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

exos_vlan

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/network/exos/exos_vlans.py [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = """
+---
+module: exos_vlans
+version_added: 2.9
```
Comment (ujwalkomarla) [inline lib/ansible/modules/network/exos/exos_vlans.py]: version_added: 2.10

File: lib/ansible/modules/network/exos/exos_vlans.py [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
+after:
+  description: The resulting configuration model invocation.
+  returned: when changed
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
+commands:
```
Comment (ujwalkomarla) [inline lib/ansible/modules/network/exos/exos_vlans.py]: Please replace `commands` section with this:
```
requests:
  description: The set of requests pushed to the remote device.
  returned: always
  type: list
  sample: [{"data": "...", "method": "...", "path": "..."}, {"data": "...", "method": "...", "path": "..."}, {"data": "...", "method": "...", "path": "..."}]
```

File: test/integration/targets/exos_vlan/defaults/main.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+---
```
Comment (ujwalkomarla) [inline test/integration/targets/exos_vlan/defaults/main.yaml]: Please move the 'exos_vlan' target folder to 'exos_vlans' folder.

File: test/integration/targets/exos_vlans/tests/httpapi/merged.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+- include_tasks: _reset_config.yaml
+
+- include_tasks: _populate_config.yaml
+
+- block:
+  - name: Merge the provided configuration with the existing running configuration
+    exos_vlans: &merged
+      config:
+        - name: vlan_30
```
Comment (ujwalkomarla) [inline test/integration/targets/exos_vlans/tests/httpapi/merged.yaml:13]: Add some vlans whose state will be suspended in the merged, overridden, replaced tests.

File: lib/ansible/module_utils/network/exos/facts/vlans/vlans.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        :param connection: the device connection
+        :param ansible_facts: Facts dictionary
+        :param data: previously collected conf
+        :rtype: dictionary
+        :returns: facts
+        """
+
+        if not data:
+            request = [{
+                "path": "/rest/restconf/data/openconfig-vlan:vlans/",
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/facts/vlans/vlans.py]: Limit the restconf GET by specifying the depth parameter with value 5 : `/rest/restconf/data/openconfig-vlan:vlans?depth=5`

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+        :rtype: A list
+        :returns: the commands necessary to remove the current configuration
+                  of the provided objects
+        """
+        requests = []
+
+        if want:
+            for w in want:
+                request_delete = deepcopy(self.VLAN_DELETE)
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: Move this line after 250.
You won't need to deepcopy unless you are actually creating a request.

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+                    h = search_obj_in_list(w['vlan_id'], have, 'vlan_id')
+                    if h:
+                        request_delete["path"] = self.DEL_PATH + str(h['vlan_id'])
+                        requests.append(request_delete)
+
+        else:
+            if not have:
+                return requests
+            for h in have:
+                request_delete = deepcopy(self.VLAN_DELETE)
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: Move this line after 261.
You won't need to deepcopy unless you are actually creating a request.

File: lib/ansible/module_utils/network/exos/argspec/vlans/vlans.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    def __init__(self, **kwargs):
+        pass
+
+    argument_spec = {
+        'config': {
+            'elements': 'dict',
+            'options': {
+                'name': {'type': 'str'},
+                'state': {
+                    'choices': ['active', 'suspended'],
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/argspec/vlans/vlans.py]: Choices are 'active' and 'suspend'.

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        """
+        requests = []
+
+        request_body = deepcopy(self.REQUEST_BODY)
+
+        for w in want:
+            if w.get('vlan_id'):
+                h = search_obj_in_list(w['vlan_id'], have, 'vlan_id')
+                if h:
+                    continue
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: If the VLAN exists, you should check the name and state to make sure they are same as requested.

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+                    have_copy.append(h)
+                else:
+                    request_post = deepcopy(self.VLAN_POST)
+                    request_body = self._update_vlan_config_body(w, request_body)
+                    request_post["data"]["openconfig-vlan:vlans"].append(request_body)
+                    request_post["data"] = json.dumps(request_post["data"])
+                    requests.append(request_post)
+
+        for h in have:
+            request_delete = deepcopy(self.VLAN_DELETE)
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: deepcopy only if required, move it inside the conditional.

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+        for w in want:
+            if w.get('vlan_id'):
+                h = search_obj_in_list(w['vlan_id'], have, 'vlan_id')
+                if h:
+                    if dict_diff(w, h):
+                        request_body = deepcopy(self.REQUEST_BODY)
+                        request_body = self._update_vlan_config_body(w, request_body)
+                        request_patch["data"]["openconfig-vlan:vlans"]["vlan"].append(request_body)
+                else:
+                    request_post = deepcopy(self.VLAN_POST)
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: You have identical code on lines 159-163, 193-197, 228-232... Move it into a function.

**Review by ujwalkomarla — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/network/exos/facts/vlans/vlans.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        :param conf: The configuration
+        :rtype: dictionary
+        :returns: The generated config
+        """
+        config = deepcopy(spec)
+
+        config["name"] = conf["name"]
+        if conf["status"] == "SUSPENDED":
+            config["state"] = "suspend"
+        else:
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/facts/vlans/vlans.py]: You can simplify with 
`config["state"] = "suspend" if conf["status"] == "SUSPENDED" else conf["status"].lower()

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+
+        result['warnings'] = warnings
+        return result
+
+    def set_config(self, existing_vlans_facts):
+        """ Collect the configuration from the args passed to the module,
+            collect the current configuration (as a dict from facts)
+
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
```
Comment (ujwalkomarla) [inline lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: If you could replace 'commands' with 'requests' in the comments that would be good.

**Review by ujwalkomarla — COMMENTED** [review]
Otherwise LGTM

**Comment by ujwalkomarla** [comment]
shipit

File: lib/ansible/module_utils/network/exos/config/vlans/vlans.py (hunk 13 above)
Comment (ujwalkomarla) [reply lib/ansible/module_utils/network/exos/config/vlans/vlans.py]: Thanks for the changes!
Looks good.

**Review by ujwalkomarla — APPROVED** [review]
shipit

**Review by ganeshrn — APPROVED** [review]
Thank you!

**Comment by ganeshrn** [comment]
rebuild_merge

**Pull request merged**